### PR TITLE
Allows Explicitly Enabling/Disabling GAX for Logging/Pubsub

### LINKED ISSUE #2521: Stackdriver ErrorReporter Unexpected Type
Hello,

The stack driver error_reporting client throws a ValueError Unexpected type.  Here is the code

```python
from google.cloud import error_reporting
error_reporter = error_reporting.Client()

try:
    raise NameError
except Exception:
    error_reporter.report_exception()
```
And the error
```python
/home/teddy/.virtualenvs/rewards/bin/python /home/teddy/PycharmProjects/rewards/workers/error_test.py
Traceback (most recent call last):
  File "/home/teddy/PycharmProjects/rewards/workers/error_test.py", line 5, in <module>
    raise NameError
NameError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/teddy/PycharmProjects/rewards/workers/error_test.py", line 7, in <module>
    error_reporter.report_exception()
  File "/home/teddy/.virtualenvs/rewards/lib/python3.5/site-packages/google/cloud/error_reporting/client.py", line 251, in report_exception
    user=user)
  File "/home/teddy/.virtualenvs/rewards/lib/python3.5/site-packages/google/cloud/error_reporting/client.py", line 183, in _send_error_report
    logger.log_struct(payload)
  File "/home/teddy/.virtualenvs/rewards/lib/python3.5/site-packages/google/cloud/logging/logger.py", line 226, in log_struct
    client.logging_api.write_entries([entry_resource])
  File "/home/teddy/.virtualenvs/rewards/lib/python3.5/site-packages/google/cloud/logging/_gax.py", line 108, in write_entries
    entry_pbs = [_log_entry_mapping_to_pb(entry) for entry in entries]
  File "/home/teddy/.virtualenvs/rewards/lib/python3.5/site-packages/google/cloud/logging/_gax.py", line 108, in <listcomp>
    entry_pbs = [_log_entry_mapping_to_pb(entry) for entry in entries]
  File "/home/teddy/.virtualenvs/rewards/lib/python3.5/site-packages/google/cloud/logging/_gax.py", line 604, in _log_entry_mapping_to_pb
    entry_pb.json_payload[key] = value
  File "/home/teddy/.virtualenvs/rewards/lib/python3.5/site-packages/google/protobuf/internal/well_known_types.py", line 732, in __setitem__
    _SetStructValue(self.fields[key], value)
  File "/home/teddy/.virtualenvs/rewards/lib/python3.5/site-packages/google/protobuf/internal/well_known_types.py", line 702, in _SetStructValue
    raise ValueError('Unexpected type')
ValueError: Unexpected type
```
It looks like a dict of `{'service': 'python'}` is getting passed to `_SetStructValue` and there is no check for a dict so it is raising a `ValueError`.

### PR DESCRIPTION
**Pull request by waprin**
Currently gax can’t log nested structs, so this is a workaround to allow forced HTTP logging.

Workaround to resolve #2521 until #2552 is resolved.

### CODE REVIEW

File: logging/google/cloud/logging/client.py [hunk 1]
```diff
@@ -76,6 +76,11 @@ class Client(JSONClient):
     _connection_class = Connection
     _logging_api = _sinks_api = _metrics_api = None
 
+    def __init__(self, project=None, credentials=None,
+                 http=None, use_gax=True):
+        super(Client, self).__init__(project, credentials, http)
+        self.use_gax = use_gax
```
Comment (dhermes) [inline logging/google/cloud/logging/client.py]: I think I'd prefer using `_USE_GAX` here. Instead let the default be `use_gax=None`, then

```python
if use_gax is None:
     self.use_gax = _USE_GAX
else:
     self.use_gax = use_gax
```

Also, any reason for `self.use_gax` to be public?

**Comment by dhermes** [comment]
Not sure this will cause coverage to go below 100% because the branches are all implicit in one-line `or` statements. But I can see the branches. So pretty please write some unit tests?

Also update the logging `Client` docstring with the new argument (pretty sure Pylint will get mad about that).

**Comment by daspecster** [comment]
What does this add that the `DISABLE_GRPC` envar doesn't do? Could the workaround just be setting that to `True`?

**Comment by dhermes** [comment]
@daspecster It's a manual override, since the gRPC client is b0rken (see the issue that @waprin linked to).

**Comment by waprin** [comment]
@daspecster Yes but error reporting currently uses the logging client, it doesn't seem right for one module to be forcing a global environment variable.

**Comment by daspecster** [comment]
Yeah I looked at the issues. If this gets added, then we may want to set a reminder/issue to take it out once the bug in gRPC is resolved right?

File: logging/google/cloud/logging/client.py (hunk 1 above)
Comment (waprin) [inline logging/google/cloud/logging/client.py]: Done, changed use_gax to be private

**Comment by waprin** [comment]
@dhermes added unit test and coverage passes.

File: logging/unit_tests/test_client.py [hunk 2]
```diff
@@ -16,7 +16,6 @@
 
 
 class TestClient(unittest.TestCase):
-
```
Comment (dhermes) [inline logging/unit_tests/test_client.py]: You delete this line for fun?

File: logging/unit_tests/test_client.py [hunk 3]
```diff
@@ -66,7 +65,6 @@ def _generated_api(*args, **kw):
             return wrapped
 
         class _GaxLoggingAPI(object):
-
```
Comment (dhermes) [inline logging/unit_tests/test_client.py]: Or maybe it's your editor being super aggressive?

File: logging/unit_tests/test_client.py [hunk 4]
```diff
@@ -475,22 +484,19 @@ def create_scoped(self, scope):
 
 
 class _DummyLoggingAPI(object):
-
```
Comment (dhermes) [inline logging/unit_tests/test_client.py]: Definitely your editor

File: logging/unit_tests/test_client.py (hunk 3 above)
Comment (waprin) [inline logging/unit_tests/test_client.py]: gah fixed

**Comment by daspecster** [comment]
@waprin looks like we have a [fix](https://github.com/GoogleCloudPlatform/google-cloud-python/issues/2552#issuecomment-254638779) to support that structure.

**Comment by dhermes** [comment]
@waprin I think a few things should happen:

1. Close this PR
1. We can have a new PR which allows `use_gax` as a constructor argument in all GAX / GAPIC / gRPC supported APIs (or re-purpose this PR to do the same)
1. We fix error reporting / logging by fixing our b0rken code

**Comment by waprin** [comment]
@dhermes SGTM, will just re-purpose this PR

**Comment by waprin** [comment]
@dhermes added pub/sub.

File: error_reporting/google/cloud/error_reporting/client.py [hunk 5]
```diff
@@ -106,7 +106,7 @@ def __init__(self, project=None,
                  service=None,
                  version=None):
         self.logging_client = google.cloud.logging.client.Client(
-            project, credentials, http)
+            project=project, credentials=credentials, http=http, use_gax=False)
```
Comment (dhermes) [inline error_reporting/google/cloud/error_reporting/client.py]: Forcing `use_gax=False` is no longer part of this PR?

File: logging/google/cloud/logging/client.py [hunk 6]
```diff
@@ -71,11 +71,23 @@ class Client(JSONClient):
     :param http: An optional HTTP object to make requests. If not passed, an
                  ``http`` object is created that is bound to the
                  ``credentials`` for the current object.
+
+    :type use_gax: bool or :class:`NoneType`
+    :param use_gax: An optional parameter that explicitly specifies whether
+                    to use the gRPC transport (gax) or HTTP
```
Comment (dhermes) [inline logging/google/cloud/logging/client.py]: This makes it seem like `gax` is the transport, instead say "(via GAX)"?

File: logging/unit_tests/test_client.py [hunk 7]
```diff
@@ -42,10 +42,10 @@ def test_logging_api_wo_gax(self):
         from google.cloud.logging.connection import _LoggingAPI
         from google.cloud.logging import client as MUT
         from google.cloud._testing import _Monkey
-        client = self._makeOne(self.PROJECT, credentials=_Credentials())
-        conn = client.connection = object()
 
         with _Monkey(MUT, _USE_GAX=False):
+            client = self._makeOne(self.PROJECT, credentials=_Credentials())
+            conn = client.connection = object()
```
Comment (dhermes) [inline logging/unit_tests/test_client.py]: Not that it matters but after this change the only thing you need in the context manager is `client = self._makeOne(...)`

File: logging/unit_tests/test_client.py [hunk 8]
```diff
@@ -85,6 +85,20 @@ def __init__(self, _wrapped):
...
         self.assertIs(again, api)
 
+    def test_no_gax_ctor(self):
+        from google.cloud.logging.connection import _LoggingAPI
+        from google.cloud.logging import client as MUT
+        from google.cloud._testing import _Monkey
+
+        creds = _Credentials()
+        with _Monkey(MUT,
+                     _USE_GAX=True):
```
Comment (dhermes) [inline logging/unit_tests/test_client.py]: Ha, need two lines here?

File: logging/unit_tests/test_client.py [hunk 9]
```diff
@@ -85,6 +85,20 @@ def __init__(self, _wrapped):
         again = client.logging_api
         self.assertIs(again, api)
 
+    def test_no_gax_ctor(self):
+        from google.cloud.logging.connection import _LoggingAPI
+        from google.cloud.logging import client as MUT
+        from google.cloud._testing import _Monkey
```
Comment (dhermes) [inline logging/unit_tests/test_client.py]: Alphabetical order? (`_testing < client < connection`)

File: logging/google/cloud/logging/client.py [hunk 10]
```diff
@@ -71,11 +71,23 @@ class Client(JSONClient):
     :param http: An optional HTTP object to make requests. If not passed, an
                  ``http`` object is created that is bound to the
                  ``credentials`` for the current object.
+
+    :type use_gax: bool or :class:`NoneType`
```
Comment (dhermes) [inline logging/google/cloud/logging/client.py]: Drop the `NoneType` (here and above)

File: logging/google/cloud/logging/client.py [hunk 11]
```diff
@@ -71,11 +71,23 @@ class Client(JSONClient):
...
+    :param use_gax: An optional parameter that explicitly specifies whether
```
Comment (dhermes) [inline logging/google/cloud/logging/client.py]: s/"An optional parameter that explicitly specifies whether"/"(Optional) Explicitly specifies whether" and then at the end of the description (i.e. after "or HTTP"), say "or HTTP. If unset, falls back to the `FOOBAR` environment variable."

File: pubsub/unit_tests/test_client.py [hunk 12]
```diff
@@ -34,10 +34,10 @@ def test_publisher_api_wo_gax(self):
         from google.cloud.pubsub import client as MUT
         from google.cloud._testing import _Monkey
         creds = _Credentials()
-        client = self._makeOne(project=self.PROJECT, credentials=creds)
-        conn = client.connection = object()
 
         with _Monkey(MUT, _USE_GAX=False):
+            client = self._makeOne(project=self.PROJECT, credentials=creds)
```
Comment (dhermes) [inline pubsub/unit_tests/test_client.py]: Ditto on fewer things in the context manager

File: pubsub/unit_tests/test_client.py [hunk 13]
```diff
@@ -84,10 +99,10 @@ def test_subscriber_api_wo_gax(self):
...
```
Comment (dhermes) [inline pubsub/unit_tests/test_client.py]: Ditto

File: error_reporting/google/cloud/error_reporting/client.py (hunk 5 above)
Comment (waprin) [inline error_reporting/google/cloud/error_reporting/client.py]: Done

File: logging/google/cloud/logging/client.py (hunk 10 above)
Comment (waprin) [inline logging/google/cloud/logging/client.py]: Done, what is the NoneType for? Saw it other places.

File: logging/google/cloud/logging/client.py (hunk 6 above)
Comment (waprin) [inline logging/google/cloud/logging/client.py]: done

File: logging/unit_tests/test_client.py (hunk 7 above)
Comment (waprin) [inline logging/unit_tests/test_client.py]: done

File: logging/unit_tests/test_client.py (hunk 9 above)
Comment (waprin) [inline logging/unit_tests/test_client.py]: done

File: logging/unit_tests/test_client.py (hunk 8 above)
Comment (waprin) [inline logging/unit_tests/test_client.py]: done

File: pubsub/unit_tests/test_client.py (hunk 12 above)
Comment (waprin) [inline pubsub/unit_tests/test_client.py]: done

File: pubsub/unit_tests/test_client.py (hunk 13 above)
Comment (waprin) [inline pubsub/unit_tests/test_client.py]: done

File: logging/google/cloud/logging/client.py (hunk 10 above)
Comment (dhermes) [inline logging/google/cloud/logging/client.py]: Well the default value is `None` and `NoneType` is the type of `None` and this field describes the type. But its overkill to say `NoneType` every time there is an optional arg.

**Comment by waprin** [comment]
@dhermes yes please squash merge at your convienence

**Pull request merged**